src/passes/ConstantFieldPropagation.cpp

```diff
@@ -0,0 +1,393 @@
+  }
+
+private:
+  FunctionStructValuesMap* functionInfos;
+
+  StructValuesMap& getInfos() { return (*functionInfos)[getFunction()]; }
+
+  // Note a value, checking whether it is a constant or not.
+  void noteExpression(Expression* expr, PossibleConstantValues& info) {
+    if (!Properties::isConstantExpression(expr)) {
+      info.noteUnknown();
+    } else {
+      info.note(Properties::getLiteral(expr));
+    }
+  }
+};
+
+// Optimize struct gets based on what we've learned about writes.
+//
+// TODO Aside from writes, we could use information like whether any struct of
+//      this type has even been created (to handle the case of struct.sets but
+//      no struct.news).
+struct FunctionOptimizer : public WalkerPass<PostWalker<FunctionOptimizer>> {
+  bool isFunctionParallel() override { return true; }
+
+  Pass* create() override { return new FunctionOptimizer(infos); }
+
+  FunctionOptimizer(StructValuesMap* infos) : infos(infos) {}
+
+  void visitStructGet(StructGet* curr) {
+    auto type = curr->ref->type;
+    if (type == Type::unreachable) {
+      return;
+    }
+
+    Builder builder(*getModule());
+
+    // Find the info for this field, and see if we can optimize. First, see if
+    // there is any information for this heap type at all. If there isn't, it is
+    // as if nothing was ever noted for that field.
+    PossibleConstantValues info;
+    assert(!info.hasNoted());
+    auto iter = infos->find(type.getHeapType());
+    if (iter != infos->end()) {
+      // There is information on this type, fetch it.
+      info = iter->second[curr->index];
+    }
+    if (!info.hasNoted()) {
+      // This field is never written at all. That means that we do not even
+      // construct any data of this type, and so it is a logic error to reach
+      // this location in the code. (Unless we are not in a closed-world
+      // situation, which we assume we are not in.) Replace this get with a
+      // trap. Note that we do not need to care about the nullability of the
+      // reference, as if it should have trapped, we are replacing it with
+      // another trap, which we allow to reorder (but we do need to care about
+      // side effects in the reference, so keep it around).
+      replaceCurrent(builder.makeSequence(builder.makeDrop(curr->ref),
+                                          builder.makeUnreachable()));
+      changedTypes = true;
+      return;
+    }
+
+    // If the value is not a constant, then it is unknown and we must give up.
+    if (!info.isConstant()) {
+      return;
+    }
+
+    // We can do this! Replace the get with a throw on a null reference (as the
+    // get would have done so), plus the constant value. (Leave it to further
+    // optimizations to get rid of the ref.)
+    replaceCurrent(builder.makeSequence(
+      builder.makeDrop(builder.makeRefAs(RefAsNonNull, curr->ref)),
+      builder.makeConstantExpression(info.getConstantValue())));
+    changedTypes = true;
```

Comment:
The struct field type and the actual value type are both known here, so we could be more precise and say the types have only changed if the types are different.

Reply:
Fair point. I'll rename the constant and add a comment. (I'd rather not add more code, which would add complexity.)